include/swift/SIL/SILBitfield.h

```diff
@@ -0,0 +1,169 @@
+//===--- SILBitField.h - Defines the bitfield utilities ---------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the BasicBlockBitfield and BasicBlockFlag utilities.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SIL_SILBITFIELD_H
+#define SWIFT_SIL_SILBITFIELD_H
+
+#include "swift/SIL/SILFunction.h"
+
+namespace swift {
+
+/// Utility to add a custom bitfield to a function's basic blocks.
+///
+/// This can be used by transforms to store temporary flags or tiny values per
+/// basic block.
+/// It is very efficient: no memory allocation is needed, no hash set or map is
+/// needed for lookup and there is no initialization cost (in contrast to
+/// BasicBlockData which needs to iterate over all blocks at initialization).
+///
+/// Restrictions:
+/// * BasicBlockBitfield instances must be allocated and deallocated
+///   following a strict stack discipline. This means, it's fine to use them as
+///   (or in) local variables in transformations. But it's e.g. not possible to
+///   store a BasicBlockBitfield in an Analysis.
+/// * The total number of bits which are alive at the same time must not exceed
+///   32.
```

Comment:
@eeckstein Post-commit review.

Can you add some more docs here? Specifically, you never describe where the actual bits are and why the bitfield needs to follow a strict bitfield discipline. I would also like a reference in the document 

```
/// Utility to add a custom bitfield to a function's basic blocks.
///
/// This can be used by transforms to store temporary flags or tiny values per
/// basic block. The memory managed is a 32 bit field within each basic block (\see BasicBlock:: customBits) and thus 
/// is very efficient: no memory allocation is needed, no hash set or map is
/// needed for lookup and there is no initialization cost (in contrast to
/// BasicBlockData which needs to iterate over all blocks at initialization).
///
/// Invariants:
/// * Each BasicBlockBitfield instance takes ownership of a range of bits from the basic block's spare
///    bits until destruction and thus those bits can not be reused by any other BasicBlockBitfield while the
///    instance is live.
/// * BasicBlockBitfield instances must be allocated and deallocated
///   following a strict stack discipline. This means, it's fine to use them as
///   (or in) local variables in transformations. But it's e.g. not possible to
///   store a BasicBlockBitfield in an Analysis.
/// * The total number of bits which are alive at the same time must not exceed
///   32 (the size of \see BasicBlock::customBits)
///
/// Some examples of utilities that use this are: ValueLifetimeAnalysis, ReachableBlocks, 
```

This would have given me all of the information that I needed to understand the code without having to search around. I found that I had to do that while

Reply:
Thanks, I improved the comment in https://github.com/apple/swift/pull/35583